src/crypto/tls/handshake_client.go

```diff
@@ -119,23 +123,65 @@ func (c *Conn) makeClientHello(minVersion uint16) (*clientHelloMsg, ecdheParamet
 	}
 
 	var params ecdheParameters
+	var keyShares []keyShare
+	var keySharePrivates []clientKeySharePrivate
+
 	if hello.supportedVersions[0] == VersionTLS13 {
 		hello.cipherSuites = append(hello.cipherSuites, defaultCipherSuitesTLS13()...)
 
-		curveID := config.curvePreferences()[0]
-		if _, ok := curveForCurveID(curveID); curveID != X25519 && !ok {
-			return nil, nil, errors.New("tls: CurvePreferences includes unsupported curve")
-		}
-		params, err = generateECDHEParameters(config.rand(), curveID)
-		if err != nil {
-			return nil, nil, err
+		if !config.AllowKEMTLS {
+			curveID := config.curvePreferences()[0]
+			if _, ok := curveForCurveID(curveID); curveID != X25519 && !ok {
+				return nil, nil, errors.New("tls: CurvePreferences includes unsupported curve")
+			}
+			params, err = generateECDHEParameters(config.rand(), curveID)
+			if err != nil {
+				return nil, nil, err
+			}
+			keyShares = append(keyShares, keyShare{group: curveID, data: params.PublicKey()})
+			keySharePrivates = append(keySharePrivates, params)
+		} else {
+			var haveECDHE, haveKEM bool
+
+			// loop over supported curves and kems until there is a KEM and an ECDHE curve
+			for _, curveID := range config.curvePreferences() {
+				if _, ok := curveForCurveID(curveID); curveID != X25519 && !curveID.isKEM() && !ok {
+					return nil, nil, errors.New("tls: CurvePreferences includes unsupported curve")
+				}
+
+				if !curveID.isKEM() && !haveECDHE {
+					params, err = generateECDHEParameters(config.rand(), curveID)
+					if err != nil {
+						return nil, nil, err
+					}
+
+					keyShares = append(keyShares, keyShare{group: curveID, data: params.PublicKey()})
+					keySharePrivates = append(keySharePrivates, params)
+					haveECDHE = true
+				} else if curveID.isKEM() && !haveKEM {
+					kemID := kem.ID(curveID)
+					pk, sk, err := kem.GenerateKey(config.rand(), kemID)
+					if err != nil {
+						return nil, nil, err
+					}
+
+					keyShares = append(keyShares, keyShare{group: curveID, data: pk.PublicKey})
+					keySharePrivates = append(keySharePrivates, sk)
+					haveKEM = true
+				}
+
+				if haveECDHE && haveKEM {
+					break
+				}
+			}
 		}
-		hello.keyShares = []keyShare{{group: curveID, data: params.PublicKey()}}
+
+		hello.keyShares = keyShares
 		hello.delegatedCredentialSupported = config.SupportDelegatedCredential
```

Comment:
What happens if `AllowKEMTLS` is set but `SupportDelegatedCredential` is not?

Reply:
There is not problem as a certificate with a kem is also allowed.

Reply:
But KEMTLS couldn't be negotiated right? Since the client isn't advertising support for DCs?

BTW, I'm not saying anything needs to be changed here. I'm just trying to understand how KEMTLS interacts with other extensions.

Reply:
If AllowKEMTLS is set that means two things:

* A KEM will be used for key exchange
* a KEM will be used as a signature: either by having it on a DC or in a leaf certificate

KEMTLS is not restricted to DCs, so if the SupportDelegatedCredential is not set to true, it is expected to fallback to a certificate that has a KEM. If a certificate with a KEM is not found, the handshake fails.